Decoupling ITM

pipeacosta opened the pull request:
This PR implements new decoupling elements based on the Ideal Transformer Model (ITM) for exemplifying co-simulation scenarios. 

This PR renames the Python API (dpsimpy) function `add` to `add_component`.

pipeacosta committed:
Implement clone method in SimNode class.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Initial implementation of decoupling node example.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement removeNode method in SystemTopology.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement new component DecouplingIdealTransformerEMT.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Add decoupling ideal transformer to the decoupling node example.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Implemented simplified decoupling node example.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Relocate simplified circuit for decoupling node testing.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement getComponents method in DecouplingIdealTransformerEMT.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Fix decoupling node example by adding ITM components.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Add extrapolation methods to the DecouplingIdealTransformerEMT class

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Improve decoupling node example with extrapolation methods

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Fixed bug in the 9 bus decoupling node example.
Modified the setParameters function of the DecouplingIdealTransformerEMT.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement command line arguments in the EMT_RC_split_decoupled_node example.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Fixed bug in the setParameters method of DecouplingIdealTransformerEMT.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Add default value to the delay parameter in EMT_RC_split_decoupled_node.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Validate co-simulation method in DecouplingIdealTransformerEMT example.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement support for zero delay in DecouplingIdealTransformerEMT.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Finalize implementation of EMT_WSCC_9bus_split_decoupled_node with zero
delay in the ITM.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Initialize co-simulation method in EMT_RC_split_decoupled_node example.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Add default prefix to EMT_RC_split_decoupled_node example.

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
- Implement voltage and current interface parameters in decoupling ITM
- Add parameters for consistent initialization using extrapolation
- Improve initialization
- Refactor: Rename decoupling node specifying that it is Ph1

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement Notebook to validate the accuracy of the decoupling ITM

Signed-off-by: Andres Acosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement new Notebook to test the decoupling ITM Ph3.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Add DecouplingIdealTransformer_EMT_Ph3 to dpsimpy

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Add extrapolation method types to dpsimpy.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement  and text DecouplingIdealTransformer_EMT_Ph3.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Improve nine bus decoupling example with 3ph decoupling ITM.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Improvements to the Decoupling ITM examples. Code cleaning.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement component DecouplingIdealTransformer_SP_Ph1.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Implement new Notebook to test the decoupling ITM.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Run pre-commit hooks.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Add internal resistors to 3-phase decoupling ITM,

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Add SP_ITM_withPiLine_Ph1.cpp to 'tests' target.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

leonardocarreras commented on dpsim-models/include/dpsim-models/Signal/DecouplingIdealTransformer_EMT_Ph1.h:
```diff
@@ -0,0 +1,111 @@
+  UInt mExtrapolationDegree = 0;
+  Eigen::MatrixXd mVoltageSrcIntfCurr;
+
+  Real interpolate(std::vector<Real> &data);
```
what does this inter/extrapolate? maybe more explanation can help

leonardocarreras commented on dpsim-models/include/dpsim-models/Signal/DecouplingIdealTransformer_EMT_Ph3.h:
```diff
@@ -0,0 +1,118 @@
+
+  class PreStep : public Task {
+  public:
+    PreStep(DecouplingIdealTransformer_EMT_Ph3 &itm)
```
this seems to be the same for both, I think a base class is also an option for a future refactoring if that is the case

leonardocarreras commented on dpsim-models/include/dpsim-models/Signal/DecouplingIdealTransformer_EMT_Ph1.h:
```diff
@@ -0,0 +1,111 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
For all the new models, please use SPDX style

leonardocarreras reviewed (approved):
Hi Andres! Thanks for the PR! This is just a quick first comment. We can discuss more and see how to integrate it.

leonardocarreras reviewed (changes requested):
Hi Andres! Please have a look to the comments! Thanks for the PR!

pipeacosta committed:
Implement DecouplingIdealTransformer_DP_Ph1 and add to pybind.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Fixes to SP ControlledCurrentSource and ITM example.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Improve ITM Notebook including the ITM in DP.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Small fixes to the ITM in EMT.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Adapt EMT_Ph3_RXLoad's clone method to use initializeParentFromNodesAndTerminals.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Clean ITM.ipynb example.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta commented on dpsim-models/include/dpsim-models/EMT/EMT_Ph3_RXLoad.h:
```diff
@@ -69,6 +71,8 @@ class RXLoad : public CompositePowerComp<Real>, public SharedFactory<RXLoad> {
   /// Initializes component from power flow data
   void initializeFromNodesAndTerminals(Real frequency) override;
+  ///
+  void initPowerFromTerminal();
```
Solved in #496

pipeacosta commented on dpsim-models/src/SystemTopology.cpp:
```diff
@@ -254,6 +254,16 @@ void SystemTopology::removeComponent(const String &name) {
+void SystemTopology::removeNode(const String &name) {
+  for (auto it = mNodes.begin(); it != mNodes.end(); ) {
+    if ((*it)->name() == name) {
+      it = mNodes.erase(it);
```
The use case we have right now, requires the user to take care of this, could we make this part of another PR?

leonardocarreras replied:
I think you can put a TODO comment there, so in the future we can implement that, and for the users not to forget to take care

pipeacosta replied:
Added TODO comment

pipeacosta committed:
Run pre-commit.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta commented on dpsim-models/include/dpsim-models/EMT/EMT_Ph3_RXLoad.h:
```diff
@@ -33,7 +33,7 @@ class RXLoad : public CompositePowerComp<Real>, public SharedFactory<RXLoad> {
   Matrix mCapacitance;
   ///
-  Bool initPowerFromTerminal = true;
+  Bool doInitPowerFromTerminal = true;
```
Solved in https://github.com/sogno-platform/dpsim/pull/496

pipeacosta commented on dpsim-models/src/EMT/EMT_Ph3_RXLoad.cpp:
```diff
@@ -72,36 +82,40 @@ void EMT::Ph3::RXLoad::setParameters(Matrix activePower, Matrix reactivePower,
-                       Logger::matrixToString(**mActivePower),
-                       Logger::matrixToString(**mReactivePower));
-    SPDLOG_LOGGER_INFO(mSLog, "Nominal Voltage={} [V]", **mNomVoltage);
+  if (doInitPowerFromTerminal) {
```
Solved in #496

pipeacosta commented on dpsim-models/src/EMT/EMT_Ph3_RXLoad.cpp:
```diff
@@ -45,7 +45,17 @@ EMT::Ph3::RXLoad::RXLoad(String name, Matrix activePower, Matrix reactivePower,
+
+SimPowerComp<Real>::Ptr EMT::Ph3::RXLoad::clone(String name) {
+  if (doInitPowerFromTerminal) {
+    initPowerFromTerminal();
```
You mean the clone function? It's because when we remove a node together with the elements connected to it, we recreate them using this function. See, for instance https://github.com/sogno-platform/dpsim/pull/423/changes#diff-b67a37da997e0b37f9461c27d7d16b0ac52ed71412ba6ccf426aee7caa6cbdc3R43

leonardocarreras replied:
I mean initPowerFromTerminal(); inside clone in this specific way

pipeacosta replied:
Fixed (removed).

pipeacosta committed:
Use logger in ITM examples.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Add explanation comments in interpolat and extrapolate functions.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Add Ideal Transformer Model (ITM) documentation.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta commented on dpsim/src/pybind/main.cpp:
```diff
@@ -235,6 +240,7 @@ PYBIND11_MODULE(dpsimpy, m) {
       .def(py::init<CPS::Real>())
       .def("add", &DPsim::SystemTopology::addComponent)
       .def("add", &DPsim::SystemTopology::addComponents)
+      .def("add_node", &DPsim::SystemTopology::addNode)
```
I decided to change the name of the one above to `add_component`, because of better semantics, but also because overloading didn't work 😅

leonardocarreras commented on dpsim-models/src/EMT/EMT_Ph3_RXLoad.cpp:
```diff
@@ -48,6 +48,16 @@ EMT::Ph3::RXLoad::RXLoad(String name, Matrix activePower, Matrix reactivePower,
 
+SimPowerComp<Real>::Ptr EMT::Ph3::RXLoad::clone(String name) {
+  if (initPowerFromTerminal) {
+    initializeParentFromNodesAndTerminals(mFrequencies(0, 0));
```
mFrequencies is not yet initialized, this call need to happen during MNA initialization for it to be valid. why do you need to call it here?

pipeacosta replied:
Fixed (removed).

leonardocarreras commented on dpsim-models/src/Signal/DecouplingIdealTransformer_DP_Ph1.cpp:
```diff
@@ -0,0 +1,199 @@
+#include "dpsim-models/Attribute.h"
+#include "dpsim-models/Definitions.h"
+#include "dpsim-models/TopologicalNode.h"
+#include <cstdlib>
```
why is this added?

pipeacosta replied:
This one is to use `std::abs`

leonardocarreras commented on dpsim-models/src/Signal/DecouplingIdealTransformer_DP_Ph1.cpp:
```diff
@@ -0,0 +1,199 @@
+#include "dpsim-models/TopologicalNode.h"
+#include <cstdlib>
+#include <dpsim-models/Signal/DecouplingIdealTransformer_DP_Ph1.h>
+#include <iostream>
```
why is this added?

pipeacosta replied:
Removed

leonardocarreras commented on dpsim-models/src/DP/DP_Ph1_ControlledVoltageSource.cpp:
```diff
@@ -0,0 +1,170 @@
+{
+  setVirtualNodeNumber(1);
+  setTerminalNumber(2);
+  **mIntfVoltage = Matrix::Zero(1, 1);
```
MatrixComp::Zero?

pipeacosta replied:
Fixed

leonardocarreras commented on dpsim-models/src/DP/DP_Ph1_ControlledVoltageSource.cpp:
```diff
@@ -0,0 +1,170 @@
+  setVirtualNodeNumber(1);
+  setTerminalNumber(2);
+  **mIntfVoltage = Matrix::Zero(1, 1);
+  **mIntfCurrent = Matrix::Zero(1, 1);
```
MatrixComp::Zero?

pipeacosta replied:
Fixed

leonardocarreras commented on dpsim-models/src/EMT/EMT_Ph3_RXLoad.cpp:
```diff
@@ -48,6 +48,16 @@ EMT::Ph3::RXLoad::RXLoad(String name, Matrix activePower, Matrix reactivePower,
+  }
+
+  auto copy = RXLoad::make(name, mLogLevel);
+  copy->setParameters(**mActivePower, **mReactivePower, **mNomVoltage);
```
`copy->setParameters(**mActivePower, **mReactivePower, **mNomVoltage, mReactanceInSeries);` might be preferable, otherwise it does not use same parameter as the original

pipeacosta replied:
Fixed

leonardocarreras commented on dpsim/examples/cxx/CIM/EMT_WSCC_9bus_split_decoupled_node.cpp:
```diff
@@ -0,0 +1,220 @@
+      "itm_" + nodeName, Logger::Level::debug);
+  idealTrafo->setParameters(nodeCopy1, nodeCopy2, 0.001, i_0, 0);
+  sys.addComponent(idealTrafo);
+  sys.addComponents(idealTrafo->getComponents());
```
`sys.addNode(idealTrafo->getVirtualNode())` might be missing after this

leonardocarreras commented on dpsim/examples/cxx/Circuits/EMT_RC_split_decoupled_node.cpp:
```diff
@@ -0,0 +1,287 @@
+  idealTrafo->setParameters(nodeCopy1, nodeCopy2, ITMDelay, irLine_0, i_inf_0,
+                            cosimMethod);
+  sys.addComponent(idealTrafo);
+  sys.addComponents(idealTrafo->getComponents());
```
`sys.addNode(idealTrafo->getVirtualNode())` might be missing

leonardocarreras commented on dpsim/examples/cxx/Circuits/EMT_RC_split_decoupled_node_DC.cpp:
```diff
@@ -0,0 +1,292 @@
+  idealTrafo->setParameters(nodeCopy1, nodeCopy2, ITMDelay, irLine_0, i_inf_0,
+                            cosimMethod);
+  sys.addComponent(idealTrafo);
+  sys.addComponents(idealTrafo->getComponents());
```
`sys.addNode(idealTrafo->getVirtualNode())` might be missing

leonardocarreras commented on dpsim-models/src/Signal/DecouplingIdealTransformer_DP_Ph1.cpp:
```diff
@@ -0,0 +1,199 @@
+  SPDLOG_LOGGER_INFO(mSLog, "Verify initial voltage: v_2 {}",
+                     mVoltageSrc->intfVoltage()(0, 0));
+
+  mCur1Extrap = std::vector<Complex>(mExtrapolationDegree + 1, 3);
```
`std::vector<Complex>(n, value)` is this intentional to have all filled with number 3 (not complex), this seems to be `Complex(0, 0)` elsewhere

pipeacosta replied:
Fixed

leonardocarreras commented on dpsim-models/src/Signal/DecouplingIdealTransformer_DP_Ph1.cpp:
```diff
@@ -0,0 +1,199 @@
+  if (mExtrapolationDegree > 0) {
+    mCur1Extrap[1] = mVoltageSrcIntfCurr(0, 0);
+  }
+  mVol2Extrap = std::vector<Complex>(mExtrapolationDegree + 1, 1);
```
`std::vector<Complex>(n, value)` is this intentional to have all filled with number 1 (not complex), this seems to be `volt2` elsewhere

pipeacosta replied:
Fixed

leonardocarreras commented on dpsim-models/src/Signal/DecouplingIdealTransformer_EMT_Ph1.cpp:
```diff
@@ -0,0 +1,179 @@
+      mSrcVoltageRef(mAttributes->create<Real>("v_ref")),
+      mSrcCurrentRef(mAttributes->create<Real>("i_ref")) {
+
+  mVoltageSrc = VoltageSource::make(name + "_v", logLevel);
```
No resistors here, like in the other cases?

pipeacosta replied:
Fixed

leonardocarreras commented on docs/hugo/content/en/docs/Models/Ideal Transformer Model/index.md:
```diff
@@ -0,0 +1,20 @@
+</figure>
+</center>
+
+To add an ITM, users must split the cirtuit and create the copies of the PCC node. An example of this process can be found in the Notebook [ITM.ipynb](../../../../../../../examples/Notebooks/Components/ITM.ipynb).
```
This would not work in Hugo, you need to point it to the repository (but then needs updating always), or just mention the notebook name.

pipeacosta replied:
Fixed

leonardocarreras commented on dpsim/src/pybind/main.cpp:
```diff
@@ -264,8 +269,9 @@ PYBIND11_MODULE(dpsimpy, m) {
-      .def("add", &DPsim::SystemTopology::addComponents)
+      .def("add_component", &DPsim::SystemTopology::addComponent)
+      .def("add_component", &DPsim::SystemTopology::addComponents)
+      .def("add_node", &DPsim::SystemTopology::addNode)
```
this needs to be mentioned in the PR text, as this breaks backwards compatibility from the Python API because the `add` method stop existing. I would rather find a way to see how to do this, as `SystemTopology` is used everywhere.

leonardocarreras reviewed (changes requested):
Hi Andres! Thanks for updating your PR!

Here some suggestions/questions:
- The ITM variants have internal virtual nodes and 4 sub-components (mRes1, mRes2, mVoltageSrc, mCurrentSrc), which is exactly what `CompositePowerComp` is designed for.  Using CompositePowerComp would also fix the missing virtual node registration bug. This could also be a future refactor.
- Please do not use the copyright block and instead use the SPDX headers.
```cpp
/* SPDX-FileCopyrightText: 2026 Institute for Automation of Complex Power Systems, EONERC, RWTH Aachen University
 * SPDX-License-Identifier: MPL-2.0
 */
```
- Breaking changes need to be explicitly mentioned (Python API case of `SystemTopology.add`)

Please feel free to propose a different way.

pipeacosta committed:
Renamed add to add_component in pybind to improve semantics.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Multiple fixes: - Use SPDX headers in new files,
- Initialize interface signals as MatrixComp instead
  of Matrix in DP_Ph1_ControlledVoltageSource
- Set correct size of extrapolation ring buffers
  in DecouplingIdealTransformer_DP_Ph1 initialization

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Remove initializeParentFromNodesAndTerminals from EMT_Ph3_RXLoad clone method.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Remove link from example in ITM documentation.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta commented on dpsim-models/include/dpsim-models/Signal/DecouplingIdealTransformer_SP_Ph1.h:
```diff
@@ -0,0 +1,116 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
Fixed

pipeacosta commented on dpsim-models/include/dpsim-models/SP/SP_Ph1_ControlledVoltageSource.h:
```diff
@@ -0,0 +1,90 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
Fixed

pipeacosta committed:
Update setParameters in EMT_Ph3_RXLoad clone method.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Added comment  to improve the removeNode function in the future.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

leonardocarreras commented on dpsim-models/src/Signal/DecouplingIdealTransformer_EMT_Ph1.cpp:
```diff
@@ -0,0 +1,188 @@
+}
+
+IdentifiedObject::List DecouplingIdealTransformer_EMT_Ph1::getComponents() {
+  return IdentifiedObject::List({mVoltageSrc, mCurrentSrc});
```
I think mRes1 and mRes2 are missing

leonardocarreras commented on dpsim/examples/cxx/Circuits/SP_ITM_withPiLine_Ph1.cpp:
```diff
@@ -0,0 +1,247 @@
+  sim.run();
+}
+
+// void simElements() {
```
please delete commented code if not used

pipeacosta committed:
Add internal resistors to DecouplingIdealTransformer_EMT_Ph1.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

pipeacosta committed:
Fix segfault bug in clone method of EMT_Ph3_RXLoad due to uninitialized power.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>

leonardocarreras commented on dpsim-models/include/dpsim-models/DP/DP_Ph1_ControlledVoltageSource.h:
```diff
@@ -0,0 +1,79 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras commented on dpsim-models/include/dpsim-models/SP/SP_Ph1_ControlledCurrentSource.h:
```diff
@@ -0,0 +1,81 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras commented on dpsim-models/src/DP/DP_Ph1_ControlledVoltageSource.cpp:
```diff
@@ -0,0 +1,170 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras commented on dpsim-models/src/SP/SP_Ph1_ControlledCurrentSource.cpp:
```diff
@@ -0,0 +1,123 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras commented on dpsim/examples/cxx/Circuits/EMT_RC_split_decoupled_node.cpp:
```diff
@@ -0,0 +1,288 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras commented on dpsim/examples/cxx/Circuits/EMT_RC_split_decoupled_node_DC.cpp:
```diff
@@ -0,0 +1,288 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras commented on dpsim/examples/cxx/Circuits/EMT_RC_split_decoupled_node_Ph3.cpp:
```diff
@@ -0,0 +1,283 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras commented on dpsim/examples/cxx/Circuits/SP_ITM_withPiLine_Ph1.cpp:
```diff
@@ -0,0 +1,169 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras commented on dpsim/examples/cxx/Circuits/SP_RC_split_decoupled_node.cpp:
```diff
@@ -0,0 +1,292 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
```
SPDX missing

leonardocarreras reviewed (changes requested):
Hi @pipeacosta things look better. I will create a follow-up issue with the pending non-blocking items. The remaining thing is SPDX-related. Could you please fix this?

pipeacosta committed:
Add missing SPDX headers.

Signed-off-by: pipeacosta <pipeacosta@gmail.com>
